### PR TITLE
Support RPC 0.4.0

### LINKED ISSUE #81: Support starknet 0.12
Rpc version 0.4.0-rc0 has just been released, we should work towards supporting it (it hits integration in ~2 weeks)
https://github.com/starkware-libs/starknet-specs/tree/v0.4.0-rc0

Let's also check if there are other (non rpc-specific things) that's changed in this release and support them.

### LINKED ISSUE #85: Change starknet_estimateMessageFee method to support v0.4.x schema
When Pathfinder will support v0.4 RPC spec we need to change the method params to match:
- https://github.com/starkware-libs/starknet-specs/pull/100/files

### LINKED ISSUE #86: Add test for starknet_estimateMessageFee method

### PR DESCRIPTION
**Pull request by DelevoXDG**
Add support for RPC 0.4.0. 
- Update transactions
- Update transaction receipts
- Add declare v0 txn
- Fix `getBlockNumber` and `getBlockHashAndNumber`
- Add tests for receipts, messages, estimateFee, estimateMessageFee - encoding tests only.
- Update `simulateTransaction` -> `simulateTransactions`
- Bump devnet to 0.6.2

Closes #81
Closes #85 
Closes #87 
Closes #86

### CODE REVIEW

File: Sources/Starknet/Data/Messages.swift [hunk 1]
```diff
@@ -1,11 +1,27 @@
...
         case payload
     }
 }
+
+public struct MessageFromL1: Codable, Equatable {
+    public let fromAddress: Felt
+    public let toAddress: Felt
+    public let selector: Felt
+    public let payload: [Felt]
+
```
Comment (dnKaratzas) [inline Sources/Starknet/Data/Messages.swift:20]: Please create a public constructor

Comment (DelevoXDG) [reply Sources/Starknet/Data/Messages.swift:20]: Do you have some sort of specific public constructor in mind?
It has a default public named constructor and I don't really think we should have a constructor with unnamed arguments for this struct.

Comment (dnKaratzas) [reply Sources/Starknet/Data/Messages.swift:20]: In order to use the struct from another module you have to provide a public constructor

File: Sources/Starknet/Data/Felt.swift [hunk 2]
```diff
@@ -41,6 +41,10 @@ public struct Felt {
     public func toHex() -> String {
         "0x\(String(value, radix: 16))"
     }
+
+    public var debugDescription: String {
+        "Felt: \(self.toHex())"
+    }
```
Comment (war-in) [inline Sources/Starknet/Data/Felt.swift:47]: Should we keep that or was it added for testing purposes only?

**Review by war-in — APPROVED** [review]
Looks good to me but @THenry14 could you please TAL?

**Comment by THenry14** [comment]
> Looks good to me but @THenry14 could you please TAL?

already looked :+1: discussed offline with @DelevoXDG , looks good to me but we'll pick this up after he's back from vacation

File: Sources/Starknet/Data/Felt.swift (hunk 2 above)
Comment (DelevoXDG) [reply Sources/Starknet/Data/Felt.swift:47]: For testing purposes, but I think we can keep it unless someone has specific reasons not to

File: Sources/Starknet/Data/Messages.swift [hunk 3]
```diff
@@ -1,11 +1,34 @@
...
 
     enum CodingKeys: String, CodingKey {
+        case fromAddress = "from_address"
         case toAddress = "to_address"
         case payload
     }
 }
+
+public struct MessageFromL1: Codable, Equatable {
+    public let fromAddress: Felt
```
Comment (bartekryba) [inline Sources/Starknet/Data/Messages.swift:16]: I see that in the spec, there's different type for ETH_ADDRESS, with fixed length to 40 digits. But I'm not sure how strict we should be about this.

File: Sources/Starknet/Data/Transaction/Transaction.swift [hunk 4]
```diff
@@ -55,7 +55,7 @@ public struct StarknetInvokeTransactionV1: StarknetTransaction {
         self.signature = try container.decode(StarknetSignature.self, forKey: .signature)
         self.maxFee = try container.decode(Felt.self, forKey: .maxFee)
         self.nonce = try container.decode(Felt.self, forKey: .nonce)
-        self.hash = try container.decode(Felt.self, forKey: .hash)
+        self.hash = try container.decodeIfPresent(Felt.self, forKey: .hash) ?? nil
```
Comment (bartekryba) [inline Sources/Starknet/Data/Transaction/Transaction.swift:58]: `?? nil` suffix is redundant here

File: Sources/Starknet/Data/Transaction/Transaction.swift [hunk 5]
```diff
@@ -110,8 +106,7 @@ public struct StarknetInvokeTransactionV0: StarknetTransaction {
         self.calldata = try container.decode(StarknetCalldata.self, forKey: .calldata)
         self.signature = try container.decode(StarknetSignature.self, forKey: .signature)
         self.maxFee = try container.decode(Felt.self, forKey: .maxFee)
-        self.nonce = try container.decode(Felt.self, forKey: .nonce)
-        self.hash = try container.decode(Felt.self, forKey: .hash)
+        self.hash = try container.decodeIfPresent(Felt.self, forKey: .hash) ?? nil
```
Comment (bartekryba) [inline Sources/Starknet/Data/Transaction/Transaction.swift:109]: Same thing here and in a few other places

**Review by bartekryba — APPROVED** [review]
Great job, lgtm.

File: Sources/Starknet/Data/Transaction/Transaction.swift (hunk 4 above)
Comment (DelevoXDG) [reply Sources/Starknet/Data/Transaction/Transaction.swift:58]: Are you sure we want to get rid of it? I know it's redundant, but makes it more obvious where nil comes from.

File: Sources/Starknet/Data/Messages.swift (hunk 3 above)
Comment (DelevoXDG) [reply Sources/Starknet/Data/Messages.swift:16]: We decided not to be strict about it in starknet.jvm, so I assume we should follow similar approach here.

File: Sources/Starknet/Data/Transaction/Transaction.swift (hunk 4 above)
Comment (bartekryba) [reply Sources/Starknet/Data/Transaction/Transaction.swift:58]: Yes

**Pull request merged**